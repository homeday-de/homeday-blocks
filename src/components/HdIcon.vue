<template>
  <inline-svg
    v-bind="$attrs"
    :transform-source="transform"
  />
</template>

<script>
import _intersection from 'lodash/intersection';
import InlineSvg from 'vue-inline-svg';

export default {
  name: 'HdIcon',
  inheritAttrs: false,
  components: {
    InlineSvg,
  },
  props: {
    rotate: {
      type: Number,
      default: 0,
    },
    fillFromClass: {
      type: Object,
      default: null,
    },
    // To add class
    classFromFill: {
      type: Object,
      default: null,
    },
  },
  methods: {
    transform(svg) {
      if (!this.fillFromClass && !this.classFromFill) {
        return svg;
      }

      const paths = svg.querySelectorAll('path');
      paths.forEach((path) => {
        const newFill = this.getFillFromClassNames(path.getAttribute('class'));
        if (newFill) {
          path.setAttribute('fill', newFill);
        }

        const newClassName = this.getClassNameFromFill(path.getAttribute('fill'));
        if (newClassName) {
          path.setAttribute('class', `${path.getAttribute('class')} ${newClassName}`);
        }
      });

      return svg;
    },
    getFillFromClassNames(classNames) {
      if (!classNames || !this.fillFromClass) {
        return null;
      }

      const classNameMatches = _intersection(
        (classNames || '').split(' '),
        Object.keys(this.fillFromClass),
      );

      return this.fillFromClass[classNameMatches[0]];
    },
    getClassNameFromFill(fill) {
      if (!fill || !this.classFromFill) {
        return '';
      }

      return this.classFromFill[fill];
    },
  },
};
</script>

<style lang="scss" scoped>

</style>
