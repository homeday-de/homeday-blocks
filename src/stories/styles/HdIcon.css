svg:hover .secondary-color {
  fill: coral;
}

b {
  font-weight: bold;
}