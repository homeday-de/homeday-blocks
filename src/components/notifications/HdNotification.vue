<template>
  <div
    class="notification"
    :class="[`notification--${type}`]"
  >
    <HdIcon
      :src="icon"
      class="notification__icon"
    />
    <div
      class="notification__content"
      :class="{
        'notification__content--centered': centered
      }"
    >
      <slot>
        {{ message }}
      </slot>
    </div>
  </div>
</template>

<script>
import HdIcon from 'homeday-blocks/src/components/HdIcon.vue';
import {
  error as errorIcon,
  bell as bellIcon,
  info as infoIcon,
  checkCircle as checkCircleIcon,
} from 'homeday-assets';

export const TYPES = {
  ERROR: 'error',
  NOTIFICATION: 'notification',
  INFO: 'info',
  SUCCESS: 'success',
};

export default {
  name: 'HdNotification',
  inheritAttrs: false,
  components: {
    HdIcon,
  },
  props: {
    type: {
      type: String,
      default: TYPES.NOTIFICATION,
      validator: type => Object.values(TYPES).includes(type),
    },
    message: {
      type: String,
      default: '',
    },
    centered: {
      type: Boolean,
      default: false,
    },
  },
  computed: {
    icon() {
      return {
        [TYPES.ERROR]: errorIcon,
        [TYPES.NOTIFICATION]: bellIcon,
        [TYPES.INFO]: infoIcon,
        [TYPES.SUCCESS]: checkCircleIcon,
      }[this.type];
    },
  },
};
</script>

<style lang="scss">
@import 'homeday-blocks/src/styles/mixins.scss';

.notification {
  display: flex;
  padding: $stack-m $inline-s;
  color: $white;

  &--notification {
    background-color: $crusta-orange;
  }

  &--success {
    background-color: $success-color;
  }

  &--error {
    background-color: $monza-red;
  }

  &--info {
    background-color: getShade($cello-blue, 80);
  }

  &__icon {
    width: 24px;
    height: 24px;
    margin-right: $inline-s;

    path {
      fill: currentColor;
    }
  }

  &__content {
    flex: 1;
    font-size: 16px;
    font-weight: 600;
    line-height: 24px;

    &--centered {
      text-align: center;
    }
  }
}
</style>
