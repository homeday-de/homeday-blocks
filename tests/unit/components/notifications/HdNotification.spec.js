import { wrapperFactoryBuilder } from 'tests/unit/helpers';
import {
  errorIcon,
  bellIcon,
  infoIcon,
  checkCircleIcon,
} from 'homeday-blocks/src/assets/small-icons';
import HdIcon from 'homeday-blocks/src/components/HdIcon.vue';
import HdNotification, { TYPES } from '@/components/notifications/HdNotification.vue';

jest.mock('homeday-blocks/src/assets/small-icons', () => ({
  errorIcon: 'errorIcon.svg',
  bellIcon: 'bellIcon.svg',
  infoIcon: 'infoIcon.svg',
  checkCircleIcon: 'checkCircleIcon.svg',
}));

const ICONS_LOOKUP_TABLE = {
  [TYPES.ERROR]: errorIcon,
  [TYPES.NOTIFICATION]: bellIcon,
  [TYPES.INFO]: infoIcon,
  [TYPES.SUCCESS]: checkCircleIcon,
};

const wrapperFactory = wrapperFactoryBuilder(HdNotification);

describe('HdNotification', () => {
  const build = (args) => {
    const wrapper = wrapperFactory({
      ...args,
      stubs: {
        HdIcon: true,
      },
    });

    return {
      wrapper,
      icon: () => wrapper.find(HdIcon),
      content: () => wrapper.find('.notification__content'),
    };
  };

  it('renders the component', () => {
    const { wrapper } = build();
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('renders all notifications types', () => {
    Object.values(TYPES).forEach((type) => {
      const expectedMessage = `Notification type: "${type}"`;
      const expectedIcon = ICONS_LOOKUP_TABLE[type];
      const expectedTypeClass = `notification--${type}`;

      const props = { message: expectedMessage, type };
      const { wrapper, icon, content } = build({ props });

      expect(icon().attributes().src).toEqual(expectedIcon);
      expect(content().text()).toContain(expectedMessage);
      expect(wrapper.classes()).toContain(expectedTypeClass);
    });
  });

  it('renders a slot message', () => {
    const expectedSlot = '<p>Hello world!, <a href="#">This is a link...</a></p>';
    const { content } = build({
      slots: {
        default: expectedSlot,
      },
    });

    expect(content().html()).toContain(expectedSlot);
  });

  it('centers content', async () => {
    const expectedClass = 'notification__content--centered';
    const { wrapper, content } = build();

    expect(content().classes()).not.toContain(expectedClass);

    wrapper.setProps({
      centered: true,
    });
    await wrapper.vm.$nextTick();

    expect(content().classes()).toContain(expectedClass);
  });
});
