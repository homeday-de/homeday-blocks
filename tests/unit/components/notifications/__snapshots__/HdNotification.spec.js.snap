// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HdNotification renders the component 1`] = `
<div class="notification notification--notification">
  <hdicon-stub src="bell.svg" class="notification__icon"></hdicon-stub>
  <div class="notification__content">

  </div>
</div>
`;
